Add support for layers and AWS credentials file

S-Cardenas opened the pull request:
I added support for [layers](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-layerversion.html#sam-layerversion-contenturi) (only local layers) specified in the AWS SAM template.

I also added the ability to specify the location of your AWS configuration and credentials file `~/.aws/credentials` so that the folder is mapped via a layer to your lambda containers and can be accessed by something like [AWS SharedInFileCredentials](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SharedIniFileCredentials.html).

S-Cardenas committed:
Publish my package to npm

S-Cardenas committed:
feat: distPath is calculated relative to cft CodeURI

S-Cardenas committed:
feat: add support for layers and the aws credentials file

S-Cardenas committed:
feat: remove sample folder from gitignore

S-Cardenas committed:
Merge pull request #1 from S-Cardenas/feature/add-support-for-layers

Feature/add support for layers

S-Cardenas committed:
feat: bump package version

S-Cardenas committed:
feat: update dockerode version to 3.3.3

S-Cardenas committed:
feat: update package version to 0.0.31

S-Cardenas committed:
Merge pull request #2 from S-Cardenas/feature-081422/update-dockerode-support-aws-lambda-nodejs-image

feat: update dockerode version to 3.3.3

S-Cardenas committed:
feat: update package-lock.json version

S-Cardenas commented:
@duartemendes @cjol @Kivol Any update on this PR? Please and thanks 👍

duartemendes commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
-  "name": "aws-sam-api-proxy",
-  "version": "0.0.20",
+  "name": "@s-cardenas/aws-sam-api-proxy",
```
We need to undo this change both here and on the `package-lock..json`.

S-Cardenas replied:
Done.

duartemendes commented on package.json:
```diff
@@ -1,6 +1,6 @@
-  "name": "aws-sam-api-proxy",
-  "version": "0.0.20",
+  "name": "@s-cardenas/aws-sam-api-proxy",
+  "version": "0.0.31",
```
Let's undo this change, the version will be updated separately from this PR along with an update on the changelog, what do you think?

S-Cardenas replied:
Agree and done.

duartemendes commented on tests/controller/index.spec.js:
```diff
@@ -23,13 +23,15 @@ describe('controller', () => {
     };
 
     controller = createController(httpClientStub, functionsWithReplicatedValue);
+    console.log('controller: ', controller);
```
Was this left here by mistake?

S-Cardenas replied:
This was a debugging mistake. I'll remove it.

duartemendes commented on src/serverlessFunctions/parseFromTemplate.js:
```diff
@@ -23,8 +23,32 @@ const buildFnPathData = (path) => {
+
+  if (prelimLayers && prelimLayers.length > 0) {
+    prelimLayers.forEach((layer) => {
+      if (!('Ref' in layer)) throw new Error('Only referenced layers are supported.');
```
Won't this be a breaking change and a blocker for people using layers without `Ref`? 🤔 

I wonder if in these cases we should simply log a warning and skip the current layer being processed, what do you think?

```suggestion
      if (!('Ref' in layer)) {
        console.warn('Only referenced layers are supported, skipping layer', layer);
        return;
      }
```

Or, do you reckon this behavior could be misleading?

S-Cardenas replied:
That's a good catch. I agree.

duartemendes commented on src/serverlessFunctions/parseFromTemplate.js:
```diff
@@ -23,8 +23,32 @@ const buildFnPathData = (path) => {
+
+      const layerName = layer.Ref;
+      const cftLayer = template.Resources[layerName];
+      const layerContentUri = cftLayer.Properties.ContentUri;
```
According to the [docs](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-layerversion.html#sam-layerversion-contenturi), `ContentUri` might be an object, which will cause the following line to break as we try to merge strings with an object.

Should we apply some validation?

S-Cardenas replied:
Also a very good catch. I will only only this layer to be processed if it's ContentUri is a relative path (i.e. string)

duartemendes commented:
> @duartemendes @cjol @Kivol Any update on this PR? Please and thanks 👍

@S-Cardenas so sorry for taking so long, unfortunately, I was not notified at the moment of the PR creation. Next time, please add me as a reviewer or tag me as you did now.

Thank you so much for your work on this, it's very appreciated.

I've left a few small comments/questions. I'm happy to merge the PR and create a new release once we go through those conversations.

Also, it would be very interesting to have the `README` file updated, briefly explaining the new feature!

S-Cardenas committed:
feat: ddd --docker-socket-path option

S-Cardenas committed:
feat: apply bug fixes for layers and contentUri

S-Cardenas committed:
feat: update readme for laers and docker socket path

S-Cardenas committed:
feat: revert name and version

S-Cardenas commented:
@duartemendes Thank you for the review. I have implemented all of your suggestions mentioned above.

Additionally I have added support for the option `--docker-socket-path` in order to allow Linux users using Docker Desktop to specify their own custom docker socket path (see the [docker docs](https://docs.docker.com/desktop/faqs/linuxfaqs/#context) for why Linux users running the desktop app have a different endpoint).

duartemendes reviewed (approved):
Thanks again for the contribution @S-Cardenas 🚀
